### PR TITLE
Fix counting of failures in PyTorch tests

### PR DESCRIPTION
**Pull request by Flamefire**
(created using `eb --new-pr`)

The RegExp handling was wrong, e.g. for an output of:

```
SKIPPED [2] distributions/test_constraints.py:83: `biject_to` not implemented.
FAILED distributions/test_constraints.py::test_constraint[True-constraint_fn5-False-value5]
FAILED distributions/test_constraints.py::test_constraint[True-constraint_fn7-True-value7]
============= 2 failed, 128 passed, 2 skipped, 2 warnings in 8.66s =============
distributions/test_constraints failed!
```

the `summary_matches_pattern2` will yield `"2 failed, 128 passed, 2 skipped, 2 warnings in 8.66s"` which does not match `r"^.*[^0-9](?Pfailures[0-9]+) failed.*$"` because there is nothing in front of the number.

This PR does a bit of refactoring to avoid this: Instead of creating `summary_matches_pattern2` out of a RegEx match and using it in 2 places way below where the exact RegEx might be already forgotten I combined the counting of errors/failures with the summary-message generation and moved the RegEx right on top of the loop handling it and added comments with examples.

Also the anchors and named group is not required:  `r"^.*[^0-9](?Pfailures[0-9]+) failed.*$"` ->  `r"([0-9]+) failed"` which is much cleaner to read.

Finally my [commit](https://github.com/easybuilders/easybuild-easyblocks/commit/30b4e02da937e319e4e6fccf4028760f2ec75b09) to also count tests terminated by a signal was lost so I readded it and reformatted the RegEx assignment to make it easier to read

And as a drive-by fix I added a docstring to `get_count_for_pattern` and made it return `0` instead of `None` for the case where the warning is shown.

### CODE REVIEW

File: easybuild/easyblocks/p/pytorch.py [hunk 1]
```diff
@@ -239,16 +239,10 @@ def configure_step(self):
...
-    def _set_cache_dir(self):
-        """Set $XDG_CACHE_HOME to avoid PyTorch defaulting to $HOME"""
-        cache_dir = os.path.join(self.tmpdir, '.cache')
-        # The path must exist!
-        mkdir(cache_dir, parents=True)
-        env.setvar('XDG_CACHE_HOME', cache_dir)
-
     def test_step(self):
         """Run unit tests"""
-        self._set_cache_dir()
```
Comment (ocaisa) [inline easybuild/easyblocks/p/pytorch.py:251]: This change looks like a regression, I'm guessing `_set_cache_dir()` was created for a reason

Comment (Flamefire) [reply easybuild/easyblocks/p/pytorch.py]: Yes, hadn't merged develop yet. Rebased and fixed and will add description now

File: easybuild/easyblocks/p/pytorch.py [hunk 2]
```diff
@@ -273,46 +273,57 @@ def test_step(self):
...
+        regex = r"^=+ (?P<failure_summary>.*) in [0-9]+\.*[0-9]*[a-zA-Z]* =+$\n(?P<failed_test_suite_name>.*) failed!$"
+
+        for summary in re.findall(regex, tests_out, re.M):
+            # E.g. '2 failed, 128 passed, 2 skipped, 2 warnings'
+            failure_summary = summary[0]
+            failure_report += "{test_suite} ({failure_summary})\n".format(
+                    test_suite=summary[1], failure_summary=failure_summary
+                )
+            failure_cnt = get_count_for_pattern(r"([0-9]+) failed", failure_summary)
+            error_cnt = get_count_for_pattern(r"([0-9]+) error", failure_summary)
```
Comment (branfosj) [inline easybuild/easyblocks/p/pytorch.py:326]:
```suggestion
            failure_cnt += get_count_for_pattern(r"([0-9]+) failed", failure_summary)
            error_cnt += get_count_for_pattern(r"([0-9]+) error", failure_summary)
```

**Review by branfosj — CHANGES_REQUESTED** [review]

**Comment by Flamefire** [comment]
Fixed yet another issue and tried to refactor it a bit to avoid more: Using `finditer` returns match objects which we can use to access RegEx groups by name.

Also in a PyTorch PR there was a syntax error in one of the tests but EB reported "Test command had non-zero exit code (1), but no failed tests found?!"

Similar if there were even a single test which failed then those syntax errors would have been undetected. I hence added my RegEx back which searches for failed test suites and if that doesn't match the ones we checked by the 2 custom regexes it will be reported as an error not taking the number of tests into account. I also include the failed test names in the error message for that case.

File: easybuild/easyblocks/p/pytorch.py [hunk 3]
```diff
@@ -266,73 +267,95 @@ def test_step(self):
...
+            failed_test_suites.append(test_suite)
+
+        # Make the names unique
+        failed_test_suites = nub(failed_test_suites)
+        # Gather all failed tests suites in case we missed any (e.g. when it exited due to syntax errors)
+        all_failed_test_suites = nub(
+            re.findall(r"^(?P<test_name>.*) failed!(?: Received signal: \w+)?\s*$", tests_out, re.M)
+        )
+        # If we missed any test suites prepend a list of all failed test suites
+        if failed_test_suites != all_failed_test_suites:
```
Comment (branfosj) [inline easybuild/easyblocks/p/pytorch.py:336]: These two will not be the same in the case where we have a failed test that matches the second pattern before one that matches the first pattern. I think we can just change the comparison to be

```py
if sorted(failed_test_suites) != sorted(all_failed_test_suites):
```

Example:

So for
```
Ran 219 tests in 67.325s

FAILED (errors=10, skipped=190, expected failures=6)
test_p1 failed!

===================== 2 failed, 128 passed, 2 skipped, 2 warnings in 3.43s =====================
test_p2 failed!
```
`failure_report` is
```
test_p1 (219 total tests, errors=10, skipped=190, expected failures=6)
test_p2 (2 failed, 128 passed, 2 skipped, 2 warnings)
```

But for
```
===================== 2 failed, 128 passed, 2 skipped, 2 warnings in 3.43s =====================
test_p2 failed!

Ran 219 tests in 67.325s

FAILED (errors=10, skipped=190, expected failures=6)
test_p1 failed!
```
`failure_report` is
```
Failed tests (suites/files):
* test_p1
* test_p2
test_p1 (219 total tests, errors=10, skipped=190, expected failures=6)
test_p2 (2 failed, 128 passed, 2 skipped, 2 warnings)
```

File: easybuild/easyblocks/p/pytorch.py [hunk 4]
```diff
@@ -266,73 +267,95 @@ def test_step(self):
...
-                msg += "{test_suite} ({failure_summary})\n".format(test_suite=summary[1], failure_summary=summary[0])
-            for summary in summary_matches:
-                msg += "{test_suite} ({total} total tests, {failure_summary})\n".format(
-                    test_suite=summary[2], total=summary[0], failure_summary=summary[1]
-                )
+            msg += failure_report
 
-            if max_failed_tests == 0:
+            # If no tests are supposed to fail or some failed for which we were not able to count errors fail now
+            if max_failed_tests == 0 or failed_test_suites != all_failed_test_suites:
```
Comment (branfosj) [inline easybuild/easyblocks/p/pytorch.py:358]: And here on comparing `failed_test_suites` to `all_failed_test_suites`.

**Review by branfosj — CHANGES_REQUESTED** [review]

File: easybuild/easyblocks/p/pytorch.py (hunk 3 above)
Comment (Flamefire) [reply easybuild/easyblocks/p/pytorch.py:336]: I guess it makes sense to not use `nub` but simply `sorted(set`. Added that as a commit.

Note that `nub` creates a set **and** a list so this might actually be faster.

**Comment by branfosj** [comment]
Test report by @branfosj
#### Overview of tested easyconfigs (in order)
 * **SUCCESS** _PyTorch-1.12.1-foss-2021a.eb_ 

Build succeeded for 1 out of 1 (1 easyconfigs in total)
bear-pg0105u36b.bear.cluster - Linux RHEL 8.5, x86_64, Intel(R) Xeon(R) Platinum 8360Y CPU @ 2.40GHz (icelake), Python 3.6.8
See https://gist.github.com/b52144157585d00a1b8d71ba060cb0a7 for a full test report.

**Comment by branfosj** [comment]
Going in, thanks @Flamefire!
